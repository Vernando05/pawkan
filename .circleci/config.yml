commands:
  create_cert:
    description: |
      Create certificates
    steps:
      - run:
          name: "Create certificates"
          command: |
            echo "$SERVERCRT" | base64 -d >> server.crt
            echo "$SERVERKEY" | base64 -d >> server.key
  build:
    description: |
      Create a production build
    steps:
      - run:
          name: "Create a production build"
          command: |
            NODE_ENV=production npm run build
            tar -cvzf build_artifacts.tar.gz dist
  test:
    description: |
      Test using Jest and Cypress
    steps:
      - run:
          name: "Jest unit testing"
          command: |
            npm run test:unit -- --collectCoverage
      - run: 
          name: "Cypress e2e testing"
          command: npm run test-ci
      - codecov/upload:
          file: coverage/clover.xml
  deploy:
    description: |
      Deploy to Firebase hosting
    steps:    
      - run: ./node_modules/.bin/firebase deploy --token "$FIREBASE_TOKEN" --only hosting
  lint:
    description: |
      Lint
    steps:
      - run:
          name: "Lint files"
          command: npm run lint
  install:
    description: |
      Install dependencies
    steps:
      - run:
          name: "Install dependencies"
          command: npm ci
orbs:
  codecov: codecov/codecov@1.1.3
executors:
  default:
    description: |
      Node 12 runtime environment
    docker:
      - image: cypress/browsers:<<parameters.cypress>>
    parameters:
      cypress:
        default: node12.13.0-chrome78-ff70
        type: string
jobs:
  build:
    description: |
      Build the application in production mode. The built assets are stored in artifacts.
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - create_cert
      - build
      - persist_to_workspace:
          paths:
            - project
          root: ~/
      - store_artifacts:
          path: ./build_artifacts.tar
  test:
    description: |
      Run the tests and generate coverage reports.
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - create_cert
      - test
  deploy:
    description: |
      Deploy to production
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - deploy
  lint:
    description: |
      Run lint on the source code
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - lint
  install:
    description: |
      Install dependencies. Checkout is called before install. (Optional) To prevent checkout from being called, pass your steps to before_install.
    executor: default
    parameters:
      before_install:
        default:
          - checkout
        description: |
          Steps to be executed before restoring cache and running install. Default is to run 'checkout' command.
        type: steps
    steps:
      - steps: <<parameters.before_install>>
      - restore_cache:
          key: v1-node-modules-{{ checksum "package-lock.json" }}
      - install
      - persist_to_workspace:
          paths:
            - project
            - .cache/Cypress
          root: ~/
      - save_cache:
          key: v1-node-modules-{{ checksum "package-lock.json" }}
          paths:
            - ~/.npm
            - ~/.cache
version: 2.1
workflows:
  version: 2.1
  front-test-build-deploy:
    jobs:
      - install
      - test:
          requires:
            - install
      - build:
          filters:
            branches:
              only:
                - dev
          requires:
            - test
      - deploy:
          filters:
            branches:
              only:
                - dev
          requires:
            - build
