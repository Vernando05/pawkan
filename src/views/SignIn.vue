<template>
  <div>
    <div v-if="errorMessages.length">
      <v-alert :key="index" v-for="(el, index) in errorMessages" type="error">{{ el.message }}</v-alert>
    </div>
    <ValidationObserver ref="observer" v-slot="{ invalid }">
      <form @submit.prevent="submit" role="form">
        <ValidationProvider v-slot="{ errors }" :name="$t('email')" rules="required|email">
          <v-text-field
            v-model="email"
            :error-messages="errors"
            :label="$t('email')"
            required
          ></v-text-field>
        </ValidationProvider>
        <ValidationProvider v-slot="{ errors }" :name="$t('password')" rules="required|min:6|max:10">
          <v-text-field
            v-model="password"
            :counter="10"
            :error-messages="errors"
            :label="$t('password')"
            type="password"
            autocomplete
            required
          ></v-text-field>
        </ValidationProvider>
        <v-btn :disabled="invalid" class="mr-4 mt-4" type="submit">{{ $t('submit') }}</v-btn>
      </form>
    </ValidationObserver>
  </div>
</template>

<script lang="ts">
  import store from '@/store'
  import mixins from 'vue-typed-mixins'
  import { mapGetters, mapMutations } from 'vuex'
  import { i18n } from '@/plugins/i18n'
  import { required, email, min, max } from 'vee-validate/dist/rules'
  import { extend, ValidationObserver, ValidationProvider, setInteractionMode } from 'vee-validate'
  import startCase from 'lodash/startCase'  
  import signIn from '@/mixins/signIn'
  import clientStorageResetToken from '@/mixins/clientStorageResetToken'
  import { CustomerUserError } from '@/types/shopify-storefront'

  setInteractionMode('aggressive')
  
  extend('required', {
    ...required,
    message: '{_field_} ' + i18n.t('cannotBeEmpty').toLowerCase()
  })

  extend('min', {
    ...min,
    message: '{_field_} ' + i18n.t('mustAtLeastHave', ['{length}']).toLowerCase()
  })

  extend('max', {
    ...max,
    message: '{_field_} ' + i18n.t('mayNotBeGreater', ['{length}']).toLowerCase()
  })

  extend('email', {
    ...email,
    message: i18n.t('emailMustBeValid'),
  })

  export default mixins(signIn, clientStorageResetToken).extend({
    name: 'SignIn',
    components: {
      ValidationProvider,
      ValidationObserver
    },
    metaInfo () {
      return {
        title: this.$t('signIn'),
        titleTemplate: '%s | Pawkan - Your best pet care'      
      }    
    },
    data: () => ({
      email: '' as string,
      password: '' as string,
      errorMessages: [] as CustomerUserError[]
    }),
    computed: {
      ...mapGetters([
        'prevNav'
      ])
    },   
    methods: { 
      ...mapMutations([
        'setCustomerAccessToken'
      ]),     
      textCapitalize (text: string): string {
        return startCase(text)
      },
      async submit (): Promise<void> {
        const isValid = await (this.$refs.observer as InstanceType<typeof ValidationProvider>).validate()
        if (isValid) {
          const result = await this.signIn_createToken({ email: this.email, password: this.password })
          if (result.customerUserErrors.length) {
            this.errorMessages = result.customerUserErrors
          } else
          if (result.customerAccessToken) {
            this.setCustomerAccessToken(result.customerAccessToken.accessToken)
            this.clientStorageResetToken_reset(result.customerAccessToken.accessToken)            
            this.email = ''
            this.password = '';
            (this.$refs.observer as InstanceType<typeof ValidationProvider>).reset()
            this.$router.push(this.prevNav)
          }
        }        
      }
    },
    beforeRouteEnter (...[, , next]) {
      if (store.getters.customerAccessToken) {
        next({ name: 'Home', params: { lang: 'en' } })
      } else {
        next()
      }
    }
  })
</script>