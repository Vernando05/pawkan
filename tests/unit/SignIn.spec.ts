import { renderComponent, flush } from './config'
import SignIn from '@/views/SignIn.vue'
import { fireEvent } from '@testing-library/vue'
import snapshotDiff  from 'snapshot-diff'

jest.useFakeTimers()

test('user filling the signin form', async () => {
  const store = {
    getters: {
      prevNav: jest.fn()
    }
  }
  const { container, getByRole, getByLabelText, queryByText } = renderComponent(SignIn, {
    store
  })
  
  await flush()
  const pristine = container.cloneNode(true)
  expect(getByRole('button', { name: /submit/i })).toBeDisabled()

  expect(queryByText(/empty/i)).not.toBeInTheDocument()
  await fireEvent.touch(getByRole('textbox', { name: /e-mail/i }))
  await flush()
  expect(queryByText(/cannot be empty/i)).toBeInTheDocument()

  await fireEvent.update(getByRole('textbox', { name: /e-mail/i }), 'jack@lumber')
  await flush()
  expect(queryByText(/empty/i)).not.toBeInTheDocument()
  expect(queryByText(/valid/i)).toBeInTheDocument()

  await fireEvent.update(getByRole('textbox', { name: /e-mail/i }), 'jack@lumber.com')
  await flush()
  expect(queryByText(/email must be valid/i)).not.toBeInTheDocument()
  expect(getByRole('button', { name: /submit/i })).toBeDisabled()

  expect(queryByText(/cannot be empty/i)).not.toBeInTheDocument()
  await fireEvent.touch(getByLabelText(/password/i))
  await flush()
  expect(queryByText(/password cannot be empty/i)).toBeInTheDocument()

  await fireEvent.update(getByLabelText(/password/i), 'ichop')
  await flush()
  expect(queryByText(/cannot be empty/i)).not.toBeInTheDocument()
  expect(queryByText(/must at least have/i)).toBeInTheDocument()

  await fireEvent.update(getByLabelText(/password/i), 'ichoptreeusingaxe')
  await flush()
  expect(queryByText(/must at least have/i)).not.toBeInTheDocument()
  expect(queryByText(/may not be greater than/i)).toBeInTheDocument()

  await fireEvent.update(getByLabelText(/password/i), 'ichoptree')
  await flush()
  expect(queryByText(/may not be greater than/i)).not.toBeInTheDocument()

  expect(getByRole('button', { name: /submit/i })).toBeEnabled()

  expect(snapshotDiff(pristine, container.cloneNode(true))).toMatchSnapshot()
})