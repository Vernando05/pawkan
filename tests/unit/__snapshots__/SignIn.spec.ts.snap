// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user filling the signin form 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -8,11 +8,11 @@
        <form
          role=\\"form\\"
        >
          <span>
            <div
-             class=\\"v-input theme--light v-text-field\\"
+             class=\\"v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted\\"
            >
              <div
                class=\\"v-input__control\\"
              >
                <div
@@ -20,11 +20,11 @@
                >
                  <div
                    class=\\"v-text-field__slot\\"
                  >
                    <label
-                     class=\\"v-label theme--light\\"
+                     class=\\"v-label v-label--active theme--light\\"
                      for=\\"input-10\\"
                      style=\\"left: 0px; position: absolute;\\"
                    >
                      E-mail
                    </label>
@@ -52,11 +52,11 @@
            </div>
          </span>
           
          <span>
            <div
-             class=\\"v-input theme--light v-text-field\\"
+             class=\\"v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted\\"
            >
              <div
                class=\\"v-input__control\\"
              >
                <div
@@ -64,11 +64,11 @@
                >
                  <div
                    class=\\"v-text-field__slot\\"
                  >
                    <label
-                     class=\\"v-label theme--light\\"
+                     class=\\"v-label v-label--active theme--light\\"
                      for=\\"input-14\\"
                      style=\\"left: 0px; position: absolute;\\"
                    >
                      Password
                    </label>
@@ -93,20 +93,19 @@
                    />
                  </div>
                  <div
                    class=\\"v-counter theme--light\\"
                  >
-                   0 / 10
+                   9 / 10
                  </div>
                </div>
              </div>
            </div>
          </span>
           
          <button
-           class=\\"mr-4 mt-4 v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default\\"
-           disabled=\\"disabled\\"
+           class=\\"mr-4 mt-4 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\"
            type=\\"submit\\"
          >
            <span
              class=\\"v-btn__content\\"
            >"
`;
